src/game-of-life/implementing.md

```diff
@@ -390,6 +390,109 @@ Make sure your development server is still running (run `npm run start` inside
 
 [![Screenshot of the Game of Life implementation with text rendering](../images/game-of-life/initial-game-of-life-pre.png)](../images/game-of-life/initial-game-of-life-pre.png)
 
+## Testing our Game of Life
+
+Now that we have our Rust implementation of the Game of Life rendering in the browser with JavaScript, let's talk about testing our Rust compiled to WebAssembly functions.
+
+We are going to test our `tick` function to make sure that it gives us the output that we expect.
+First, we'll want to create some setter functions inside our `impl Universe` in the `wasm_game_of_life/src/lib.rs` file. We are going to create a `set_width` and a `set_height` function so we can create `Universe`s of different sizes.
+
+```rust
+#[wasm_bindgen]
+impl Universe { 
+    // ...
+
+    pub fn set_width(mut self, width: u32) -> Self {
+        self.width = width;
+        // Reinitialize cells with new width and the universe's height.
+        // Set all the cells to dead.
+        self.cells = (0..width * self.height).map(|_i| Cell::Dead).collect();
+        self
+    }
+
+    pub fn set_height(mut self, height: u32) -> Self {
+        self.height = height;
+        // Reinitialize cells with new height and the universe's width.
+        // Set all the cells to dead.
+        self.cells = (0..self.width * height).map(|_i| Cell::Dead).collect();
+        self
+    }
+
+}
+```
+
+Now we are going to make some public Rust functions for our `Universe`. We aren't going to compile them to WebAssembly, however. Rust compiled into WebAssembly functions cannot return borrowed references or take dynamically sized vectors as function arguments. Try compiling the Rust to WebAssembly by adding the `#[wasm_bindgen]` attribute and take a look at the errors you get.
+
+```rust
+pub fn cells(u: &Universe) -> &[Cell] {
+    &u.cells
+}
+
+/// Pass a vector of indices to set only those cells to be alive.
+pub fn set_cells(u: &mut Universe, indices: Vec<u32>) -> &mut Universe {
```

Comment:
A few things:

* I don't think this needs to return the universe, does it?

* Since these functions are only used for tests, we can get them on whether we are compiling for testing or not by adding `#[cfg(test)]` to them.

* Rather than taking indices, and making the authors of tests go from row+col to index by hand, we can take row+col pairs in `set_cells` and use our extant `get_index` function.

* I think `set_cells` and `cells` would be a little cleaner as methods of `Universe`.

Putting all that together, we get something like this:

```rust
// Note: no `#[wasm_bindgen]` on this impl block, since these methods are
// not intended to be exposed to JS.
//
// Note 2: These are testing-only methods, and are not compiled when creating
// normal builds because of the `cfg` attribute.
#[cfg(test)]
impl Universe {
    pub fn get_cells(&self) -> &[Cell] {
        ...
    }

    pub fn set_cells(&mut self, cells: &[(u32, u32)]) {
        ...
    }
}
```

Reply:
I'm a little confused about adding `#[cfg(test)]` & removing `#[wasm_bindgen]`. I thought the suggestion might've been to create a new `impl Universe` block in the `wasm-game-of-life/tests/web.rs` file & have the `#[cfg(test)]` attribute there. That didn't work since it overlaps with the `wasm_game_of_life` definition. 

I'm also confused about removing the `#[wasm_bindgen]` attribute if it's the same `impl Universe` block in `lib.rs`. I thought the point was to test the Rust-generated WebAssembly `tick` function. However, with the new changes I've pushed, it doesn't seem like I'm generating any of the methods on `Universe` as WebAssembly. What am I missing here?

I think I managed to get the rest of the changes you suggested in the other comments. I imagine adding more explanation & possibly more examples would be good, but I think this is a step in the right direction.

Thank you for your detailed feedback. I appreciate being pointed towards writing more idiomatic Rust rather than the "Haskell-styled" Rust I'm use to writing. :)

Reply:
Yes, I was suggesting a new `impl` block, but in `lib.rs` not the test file. This new `impl` block would have `cfg(test)` so that it didn't get built for non-tests despite being in the main file. 

> That didn't work since it overlaps with the `wasm_game_of_life` definition.

Do you mean there are conflicting method names? If so, we can prefix our testing-only methods with `testing_only_*`.

> I'm also confused about removing the `#[wasm_bindgen]` attribute if it's the same `impl Universe` block in `lib.rs`.

Different `impl` block :) The extant one that exposes methods to JS should remain the same.

> I imagine adding more explanation & possibly more examples would be good, but I think this is a step in the right direction.

For sure -- I'm a big fan of landing incremental, small-but-complete changes and then doing more work in follow ups.

> Thank you for your detailed feedback. I appreciate being pointed towards writing more idiomatic Rust rather than the "Haskell-styled" Rust I'm used to writing. :)

Great :)

Let me know if anything still isn't clear; looking forward to landing this!